straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,316 @@ def find_n_competing(peaks, windows, fraction):
+            split_peaks = strax.touching_windows(peaks[mask_pre], roi_shadow)
+            array = np.zeros(len(current_peak), np.dtype(self.shadowdtype))
+
+            # Initialization
+            array['pre_x'] = np.nan
+            array['pre_y'] = np.nan
+            array['pre_dt'] = self.time_window_backward
+            # Shadow is set to be the lowest possible value
+            if 'time' in key:
+                array['shadow'] = self.threshold[key] * array['pre_dt'] ** self.exponent
+            else:
+                array['shadow'] = 0
+            array['nearest_dt'] = self.time_window_backward
+
+            # Calculating shadow, the Major of the plugin. Only record the previous peak casting the largest shadow
+            if len(current_peak):
+                self.peaks_shadow(current_peak, peaks[mask_pre], split_peaks, self.exponent, array, is_position, self.getsigma(self.sigma, self.baseline, current_peak['area']))
+            
+            # Fill results
+            names = ['shadow', 'pre_dt']
+            if 's2' in key: # Only previous S2 peaks have (x,y)
+                names += ['pre_x', 'pre_y']
+            if 'time' in key: # Only time shadow gives the nearest large peak
+                names += ['nearest_dt']
+            for name in names:
+                if name == 'nearest_dt':
+                    result[f'{name}_{type_str}'] = array[name]
+                else:
+                    result[f'{name}_{key}'] = array[name]
+
+        distance = np.sqrt((result[f'pre_x_s2_position'] - current_peak['x']) ** 2 + (result[f'pre_y_s2_position'] - current_peak['y']) ** 2)
+        # If distance is NaN, set largest distance
+        distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+        # HalfCauchy PDF when calculating S2 position shadow
+        result['shadow_s2_position_pdf'] = halfcauchy.pdf(distance, scale=self.getsigma(self.sigma, self.baseline, current_peak['area']))
+
+        # 6. Set time and endtime for peaks
+        result['time'] = current_peak['time']
+        result['endtime'] = strax.endtime(current_peak)
+        return result
+
+    @staticmethod
+    def getsigma(sigma, baseline, s2):
+        # The parameter of HalfCauchy, which is a function of S2 area
+        return sigma / np.sqrt(s2) + baseline
 
     @staticmethod
     @numba.njit
-    def compute_shadow(peaks, pre_s2_peaks, touching_windows, exponent, res):
+    def peaks_shadow(peaks, pre_peaks, touching_windows, exponent, result, pos_corr, sigmas=None):
         """
-        For each peak in peaks, check if there is a shadow-casting S2 peak
+        For each peak in peaks, check if there is a shadow-casting peak
         and check if it casts the largest shadow
         """
+        for p_i, (p_a, sigma) in enumerate(zip(peaks, sigmas)):
+            indices = touching_windows[p_i]
+            for idx in range(indices[0], indices[1]):
+                s_a = pre_peaks[idx]
```

Comment:
Can we rename s_a in something more readable like `shadow_peak` or `shadow_casting_peak`?

Reply:
Same for p_a ? I do not know what this `a` stands for.